test/unit/test_str_search.c

```diff
@@ -0,0 +1,189 @@
+
+	RzList *str_list = rz_list_new();
+	int n = rz_scan_strings(buf, str_list, &g_opt, 0, buf->methods->get_size(buf) - 1, RZ_STRING_ENC_GUESS);
+	mu_assert_eq(n, 1, "rz_scan_strings utf32le, number of strings");
+
+	RzDetectedString *s = rz_list_get_n(str_list, 0);
+	mu_assert_streq(s->string, "I am a UTF-32le string", "rz_scan_strings utf32le, different string");
+	mu_assert_eq(s->addr, 2, "rz_scan_strings utf32le, address");
+	mu_assert_eq(s->type, RZ_STRING_ENC_UTF32LE, "rz_scan_strings utf32le, string type");
+
+	rz_detected_string_free(s);
+	rz_list_free(str_list);
+
+	mu_end;
+}
+
+bool test_rz_scan_strings_detect_utf32_be(void) {
+	static const unsigned char str[] =
+		"\xff\xff\x00\x00\x00\x49\x00\x00\x00\x20\x00\x00\x00\x61\x00"
+		"\x00\x00\x6d\x00\x00\x00\x20\x00\x00\x00\x61\x00\x00\x00\x20"
+		"\x00\x00\x00\x55\x00\x00\x00\x54\x00\x00\x00\x46\x00\x00\x00"
+		"\x2d\x00\x00\x00\x33\x00\x00\x00\x32\x00\x00\x00\x62\x00\x00"
+		"\x00\x65\x00\x00\x00\x20\x00\x00\x00\x73\x00\x00\x00\x74\x00"
+		"\x00\x00\x72\x00\x00\x00\x69\x00\x00\x00\x6e\x00\x00\x00\x67"
+		"\x00\x00\x00\xff\xff\xff\xff\xff\xff\xff\xff";
+
+	RzBuffer *buf = rz_buf_new_with_bytes(str, sizeof(str));
+
+	RzList *str_list = rz_list_new();
+	int n = rz_scan_strings(buf, str_list, &g_opt, 0, buf->methods->get_size(buf) - 1, RZ_STRING_ENC_GUESS);
+	mu_assert_eq(n, 1, "rz_scan_strings utf32be, number of strings");
+
+	RzDetectedString *s = rz_list_get_n(str_list, 0);
+	mu_assert_streq(s->string, "I am a UTF-32be string", "rz_scan_strings utf32be, different string");
+	mu_assert_eq(s->addr, 2, "rz_scan_strings utf32be, address");
+	mu_assert_eq(s->type, RZ_STRING_ENC_UTF32BE, "rz_scan_strings utf32be, string type");
+
+	rz_detected_string_free(s);
+	rz_list_free(str_list);
+
+	mu_end;
+}
+
+bool test_rz_scan_strings_utf16_be(void) {
+	static const unsigned char str[] =
+		"\xff\xfftorre, alfiere\xff\x00\x04\x41\x04\x3b\x04\x3e\x04\x3d\x00\x2c\x00\x20\x04\x3b\x04\x30\x04\x34\x04\x4c\x04\x4f";
+
+	RzBuffer *buf = rz_buf_new_with_bytes(str, sizeof(str));
+
+	RzList *str_list = rz_list_new();
+	int n = rz_scan_strings(buf, str_list, &g_opt, 16, buf->methods->get_size(buf) - 1, RZ_STRING_ENC_UTF16BE);
+	mu_assert_eq(n, 1, "rz_scan_strings utf16be, number of strings");
+
+	RzDetectedString *s = rz_list_get_n(str_list, 0);
+
+	mu_assert_eq(s->addr, 18, "rz_scan_strings utf16be, address");
+	mu_assert_streq(s->string, "\xd1\x81\xd0\xbb\xd0\xbe\xd0\xbd\x2c\x20\xd0\xbb\xd0\xb0\xd0\xb4\xd1\x8c\xd1\x8f",
+		"rz_scan_strings utf16be, different string");
+	mu_assert_eq(s->type, RZ_STRING_ENC_UTF16BE, "rz_scan_strings utf16be, string type");
+
+	rz_detected_string_free(s);
+	rz_list_free(str_list);
+
+	mu_end;
+}
+
+bool all_tests() {
+	mu_run_test(test_rz_scan_strings_detect_ascii);
+	mu_run_test(test_rz_scan_strings_detect_utf8);
+	mu_run_test(test_rz_scan_strings_detect_utf16_le);
+	// mu_run_test(test_rz_scan_strings_detect_utf16_be);
```

Comment:
Why these are commented out?

Reply:
Because the buffers are ambiguous. For example, the UTF16 string can be either
`I am a UTF-16be string` UTF16-be starting at offset 3
or
`I am a UTF-16be strin` UTF16-le starting at offset 4

In the current state of the PR, the string is detected as the second one. I would like to change the test so that the buffer is non-ambiguous.

Reply:
With my suggestion you could set the preferred option in these test to avoid ambiguity. I also recommend to add a couple more tests with the different preferred string endianness in the options.